### DISCUSSION TITLE
Symfony semantic versioning and PHP requirements

### DISCUSSION THREAD

**Discussion by exteon** [symfony/symfony#46939] [Q&A] [answered]
Today my symfony `^6.0` projects got broken by `6.1` requiring `PHP 8.1`
I couldn't find it in the docs, but how exactly is Symfony versioned? Should I use `~6.0.0` instead?

Note: I'm not using the fiull symfony stack, I'm using `config` here and there, `yaml` here and there, `console` here and everywhere... etc
Today the break was triggered by `symfony/config[6.1]`

**Comment by norkunas** [comment]
There was a blog post about that on February: https://symfony.com/blog/symfony-6-1-will-require-php-8-1

**Comment by exteon** [comment]
Thanks @norkunas, so if I read the blog post in the right key, the answer to my question would be:

1. This is a rather a singular occurence
2. ALL _(Hm, any dependable exceptions here?)_ `symfony/* ` lib deps should be converted from `^6.0` to `~6.0.0`
3. This is also a [composer](https://github.com/composer/composer) limitation ... It would be so incredibly nice if composer would be able to retrofit environment constraints, such as `(symfony/symfony ^6.0) && (PHP =8.0) => symfony/symfony ~6.0`, like ruby (and I think python) package managers are able to.

Is this about right?

**Reply by stof** [reply]
What do you mean by "my project was broken" ? Composer takes the PHP requirements into account when resolving dependencies. This is the case since the first alpha version of Composer. The behavior you describe in your point 3 is already the behavior of composer.
could it be that you resolved your dependencies using a PHP 8.1 runtime ?

**Reply by exteon** [reply]
> What do you mean by "my project was broken" ? 
> could it be that you resolved your dependencies using a PHP 8.1 runtime ?

Nope, PHP 8.0 all around.
I started off with an older project and I did a `compose require my/lib` where `my/lib` did not directly have symfony dependencies, if any co-dependencies existed they were all specified as `^` constraints...
This and any subsequent attempt to `composer update` would throw something like "my/x requires symfony/config ^6.0, satisfiable by symfony/config[6.1], bla bla PHP 8.1"... update only resumed working after I modified composer dependencies to `~` constraint.

I never had the pleasure of dealing with this issue before, so if this looks like a bug rather than intended behavior, I'll try to reproduce a minimal case and forward to either composer or symfony where appropriate.

**Reply by exteon** [reply]
If it's a bug, I think it might have been set off by `composer require` which I think always assumes only the latest version for a library... hm... I just come to think of it. I'll try tto retrace my steps. But either way, subsequent `composer update` also failed with no dependency on any >=6.1 symfony libs whatsoever. It just went for `symfony/config` 6.1 so I assumed composer dowsn't retrofit environment constraints, it just validates them at the end.

**Reply by exteon** [reply]
Ok, I just changed all dependencies in all libraries back to `^6.0.0` and `composer update` works. So, definitely composer-related, we can close this here. I'll try to recreate it and post to composer if I succeed.

**Comment by exteon** [comment, marked as answer]
Ok, I just changed all dependencies in all libraries back to ^6.0.0 and composer update works. So I guess any bug, if it exists, is with composer.
